File: rustspider/src/performance/limiter.rs
//! 性能优化模块
//! 包含速率限制、熔断器、连接池等

use parking_lot::Mutex;
use std::collections::hash_map::DefaultHasher;
use std::collections::HashMap;
use std::hash::{Hash, Hasher};
use std::sync::atomic::{AtomicUsize, Ordering};
use std::sync::Arc;
use std::thread;
use std::time::{Duration, Instant};

/// 速率限制器（令牌桶算法）
pub struct RateLimiter {
    rate: usize,
    interval: Duration,
    tokens: AtomicUsize,
    last_refill: Mutex<Instant>,
}

impl RateLimiter {
    /// 创建速率限制器
    pub fn new(rate: usize, interval_secs: u64) -> Self {
        RateLimiter {
            rate,
            interval: Duration::from_secs(interval_secs),
            tokens: AtomicUsize::new(rate),
            last_refill: Mutex::new(Instant::now()),
        }
    }

    /// 等待获取令牌
    pub fn wait(&self) {
        loop {
            self.refill();

            let current = self.tokens.load(Ordering::SeqCst);
            if current > 0 {
                if self
                    .tokens
                    .compare_exchange(current, current - 1, Ordering::SeqCst, Ordering::SeqCst)
                    .is_ok()
                {
                    break;
                }
            } else {
                thread::sleep(Duration::from_millis(100));
            }
        }
    }

    fn refill(&self) {
        let mut last_refill = self.last_refill.lock();
        let now = Instant::now();
        let elapsed = now.duration_since(*last_refill);

        let tokens_to_add =
            (elapsed.as_secs_f64() / self.interval.as_secs_f64()) as usize * self.rate;

        if tokens_to_add > 0 {
            let current = self.tokens.load(Ordering::SeqCst);
            self.tokens.store(
                std::cmp::min(self.rate, current + tokens_to_add),
                Ordering::SeqCst,
            );
            *last_refill = now;
        }
    }
}

/// 熔断器状态
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum CircuitState {
    Closed,
    Open,
    HalfOpen,
}

/// 熔断器
pub struct CircuitBreaker {
    failure_threshold: usize,
    success_threshold: usize,
    timeout: Duration,

    failures: AtomicUsize,
    successes: AtomicUsize,
    state: Mutex<CircuitState>,
    last_failure: Mutex<Option<Instant>>,
}

impl CircuitBreaker {
    /// 创建熔断器
    pub fn new(failure_threshold: usize, success_threshold: usize, timeout_secs: u64) -> Self {
        CircuitBreaker {
            failure_threshold,
            success_threshold,
            timeout: Duration::from_secs(timeout_secs),
            failures: AtomicUsize::new(0),
            successes: AtomicUsize::new(0),
            state: Mutex::new(CircuitState::Closed),
            last_failure: Mutex::new(None),
        }
    }

    /// 检查是否允许请求
    pub fn allow(&self) -> bool {
        let mut state = self.state.lock();

        match *state {
            CircuitState::Open => {
                if let Some(last_failure) = *self.last_failure.lock() {
                    if last_failure.elapsed() > self.timeout {
                        *state = CircuitState::HalfOpen;
                        return true;
                    }
                }
                false
            }
            _ => true,
        }
    }

    /// 记录成功
    pub fn record_success(&self) {
        let mut state = self.state.lock();

        if *state == CircuitState::HalfOpen {
            let successes = self.successes.fetch_add(1, Ordering::SeqCst) + 1;
            if successes >= self.success_threshold {
                *state = CircuitState::Closed;
                self.failures.store(0, Ordering::SeqCst);
                self.successes.store(0, Ordering::SeqCst);
            }
        } else if *state == CircuitState::Closed {
            self.failures.store(0, Ordering::SeqCst);
        }
    }

    /// 记录失败
    pub fn record_failure(&self) {
        let mut state = self.state.lock();
        let failures = self.failures.fetch_add(1, Ordering::SeqCst) + 1;

        *self.last_failure.lock() = Some(Instant::now());

        if *state == CircuitState::HalfOpen || failures >= self.failure_threshold {
            *state = CircuitState::Open;
            self.successes.store(0, Ordering::SeqCst);
        }
    }

    /// 获取状态
    pub fn state(&self) -> CircuitState {
        *self.state.lock()
    }
}

/// 连接池
pub struct ConnectionPool {
    max_connections: usize,
    current: AtomicUsize,
    semaphore: Arc<Semaphore>,
}

impl ConnectionPool {
    /// 创建连接池
    pub fn new(max_connections: usize) -> Self {
        ConnectionPool {
            max_connections,
            current: AtomicUsize::new(0),
            semaphore: Arc::new(Semaphore::new(max_connections)),
        }
    }

    /// 获取连接
    pub fn acquire(&self) -> Option<ConnectionPermit<'_>> {
        if self.semaphore.acquire() {
            self.current.fetch_add(1, Ordering::SeqCst);
            Some(ConnectionPermit { pool: self })
        } else {
            None
        }
    }

    /// 释放连接
    fn release(&self) {
        self.current.fetch_sub(1, Ordering::SeqCst);
        self.semaphore.release();
    }

    /// 获取统计
    pub fn stats(&self) -> ConnectionStats {
        ConnectionStats {
            current: self.current.load(Ordering::SeqCst),
            max: self.max_connections,
            available: self.max_connections - self.current.load(Ordering::SeqCst),
        }
    }
}

/// 连接许可
pub struct ConnectionPermit<'a> {
    pool: &'a ConnectionPool,
}

impl<'a> Drop for ConnectionPermit<'a> {
    fn drop(&mut self) {
        self.pool.release();
    }
}

/// 连接统计
pub struct ConnectionStats {
    pub current: usize,
    pub max: usize,
    pub available: usize,
}

/// 信号量
struct Semaphore {
    count: AtomicUsize,
}

impl Semaphore {
    fn new(count: usize) -> Self {
        Semaphore {
            count: AtomicUsize::new(count),
        }
    }

    fn acquire(&self) -> bool {
        loop {
            let current = self.count.load(Ordering::SeqCst);
            if current == 0 {
                return false;
            }
            if self
                .count
                .compare_exchange(current, current - 1, Ordering::SeqCst, Ordering::SeqCst)
                .is_ok()
            {
                return true;
            }
        }
    }

    fn release(&self) {
        self.count.fetch_add(1, Ordering::SeqCst);
    }
}

/// 自适应速率限制器
pub struct AdaptiveRateLimiter {
    initial_delay: f64,
    min_delay: f64,
    max_delay: f64,
    target_response_time: f64,
    domain_delays: Mutex<HashMap<String, f64>>,
    domain_last_request: Mutex<HashMap<String, f64>>,
}

impl AdaptiveRateLimiter {
    /// 创建自适应速率限制器
    pub fn new(
        initial_delay: f64,
        min_delay: f64,
        max_delay: f64,
        target_response_time: f64,
    ) -> Self {
        AdaptiveRateLimiter {
            initial_delay,
            min_delay,
            max_delay,
            target_response_time,
            domain_delays: Mutex::new(HashMap::new()),
            domain_last_request: Mutex::new(HashMap::new()),
        }
    }

    /// 等待
    pub fn wait(&self, url: &str) {
        let domain = self.extract_domain(url);

        let delays = self.domain_delays.lock();
        let mut last_requests = self.domain_last_request.lock();

        let delay = delays.get(&domain).copied().unwrap_or(self.initial_delay);
        let last_request = last_requests.get(&domain).copied().unwrap_or(0.0);

        let now = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_secs_f64();

        let elapsed = now - last_request;
        if elapsed < delay {
            thread::sleep(Duration::from_secs_f64(delay - elapsed));
        }

        last_requests.insert(domain.clone(), now);
    }

    /// 调整延迟
    pub fn adjust(&self, url: &str, response_time: f64, status_code: u16) {
        let domain = self.extract_domain(url);

        let mut delays = self.domain_delays.lock();
        let current_delay = delays.get(&domain).copied().unwrap_or(self.initial_delay);
        let mut new_delay = current_delay;

        // 根据响应时间调整
        if response_time < self.target_response_time {
            new_delay = current_delay * 0.9;
        } else if response_time > self.target_response_time * 2.0 {
            new_delay = current_delay * 1.5;
        }

        // 根据状态码调整
        match status_code {
            429 | 503 => new_delay = current_delay * 2.0,
            500..=599 => new_delay = current_delay * 1.2,
            _ => {}
        }

        // 应用限制
        new_delay = new_delay.max(self.min_delay).min(self.max_delay);

        if (new_delay - current_delay).abs() > 0.1 {
            delays.insert(domain, new_delay);
        }
    }

    fn extract_domain(&self, url: &str) -> String {
        if let Ok(parsed) = url::Url::parse(url) {
            parsed.host_str().unwrap_or("").to_string()
        } else {
            String::new()
        }
    }
}

/// 内容指纹（去重）
pub struct ContentFingerprinter {
    seen_hashes: Mutex<std::collections::HashSet<u64>>,
}

impl ContentFingerprinter {
    /// 创建内容指纹
    pub fn new() -> Self {
        ContentFingerprinter {
            seen_hashes: Mutex::new(std::collections::HashSet::new()),
        }
    }

    /// 检查是否重复
    pub fn is_duplicate(&self, content: &str) -> bool {
        let hash = self.hash(content);
        let mut seen = self.seen_hashes.lock();

        if seen.contains(&hash) {
            true
        } else {
            seen.insert(hash);
            false
        }
    }

    fn hash(&self, content: &str) -> u64 {
        let mut hasher = DefaultHasher::new();
        content.hash(&mut hasher);
        hasher.finish()
    }

    /// 清空
    pub fn clear(&self) {
        self.seen_hashes.lock().clear();
    }
}

impl Default for ContentFingerprinter {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_rate_limiter() {
        let limiter = RateLimiter::new(5, 1);
        limiter.wait();
        limiter.wait();
    }

    #[test]
    fn test_circuit_breaker() {
        let cb = CircuitBreaker::new(3, 2, 60);

        assert!(cb.allow());
        cb.record_failure();
        cb.record_failure();
        cb.record_failure();

        assert!(!cb.allow()); // 应该打开
    }

    #[test]
    fn test_fingerprinter() {
        let fp = ContentFingerprinter::new();

        assert!(!fp.is_duplicate("test"));
        assert!(fp.is_duplicate("test"));
    }
}

File: pyspider/performance/limiter.py
"""
性能优化模块
包含速率限制、熔断器、连接池等
"""

import time
import threading
from typing import Dict, Optional
from datetime import datetime


class RateLimiter:
    """速率限制器（令牌桶算法）"""

    def __init__(self, rate: int, interval: float = 1.0):
        """
        Args:
            rate: 每秒请求数
            interval: 时间间隔（秒）
        """
        self.rate = rate
        self.interval = interval
        self.capacity = float(rate)
        self.tokens = float(rate)
        self.last_refill = time.monotonic()
        self._lock = threading.Lock()

    def acquire(self):
        """获取令牌"""
        while True:
            with self._lock:
                self._refill()
                if self.tokens >= 1:
                    self.tokens -= 1
                    return
                wait_time = max((1 - self.tokens) / float(self.rate or 1), 0.01)
            time.sleep(wait_time)

    def wait(self):
        """兼容旧接口"""
        self.acquire()

    def _refill(self):
        """补充令牌"""
        now = time.monotonic()
        elapsed = now - self.last_refill
        tokens_to_add = elapsed * (float(self.rate) / float(self.interval or 1))
        if tokens_to_add > 0:
            self.tokens = min(self.capacity, self.tokens + tokens_to_add)
            self.last_refill = now


class TokenBucket:
    """通用令牌桶，替代旧的 spider_async_v3 实现。"""

    def __init__(
        self,
        rate: Optional[float] = None,
        capacity: Optional[float] = None,
        refill_rate: Optional[float] = None,
    ):
        resolved_refill_rate = (
            refill_rate
            if refill_rate is not None
            else (rate if rate is not None else 1.0)
        )
        resolved_capacity = capacity if capacity is not None else resolved_refill_rate

        self.capacity = float(resolved_capacity)
        self.refill_rate = float(resolved_refill_rate)
        self.rate = self.refill_rate
        self.tokens = float(resolved_capacity)
        self.last_update = time.monotonic()
        self._lock = threading.Lock()

    def _refill(self):
        now = time.monotonic()
        elapsed = now - self.last_update
        self.tokens = min(self.capacity, self.tokens + elapsed * self.refill_rate)
        self.last_update = now

    def consume(self, amount: float = 1) -> bool:
        with self._lock:
            self._refill()
            if self.tokens >= amount:
                self.tokens -= amount
                return True
            return False


class CircuitBreaker:
    """熔断器"""

    def __init__(
        self,
        failure_threshold: int = 5,
        success_threshold: int = 3,
        timeout: float = 60.0,
    ):
        self.failure_threshold = failure_threshold
        self.success_threshold = success_threshold
        self.timeout = timeout

        self.failures = 0
        self.successes = 0
        self.state = "closed"  # closed, open, half-open
        self.last_failure: Optional[datetime] = None
        self._lock = threading.Lock()

    def allow(self) -> bool:
        """检查是否允许请求"""
        with self._lock:
            if self.state == "open":
                if (
                    self.last_failure
                    and (datetime.now() - self.last_failure).total_seconds()
                    > self.timeout
                ):
                    self.state = "half-open"
                    return True
                return False
            return True

    def record_success(self):
        """记录成功"""
        with self._lock:
            if self.state == "half-open":
                self.successes += 1
                if self.successes >= self.success_threshold:
                    self.state = "closed"
                    self.failures = 0
                    self.successes = 0
            elif self.state == "closed":
                self.failures = 0

    def record_failure(self):
        """记录失败"""
        with self._lock:
            self.failures += 1
            self.last_failure = datetime.now()

            if self.state == "half-open" or self.failures >= self.failure_threshold:
                self.state = "open"
                self.successes = 0

    def get_state(self) -> str:
        """获取状态"""
        with self._lock:
            return self.state


class ConnectionPool:
    """连接池"""

    def __init__(self, max_connections: int = 100):
        self.max_connections = max_connections
        self.current = 0
        self._lock = threading.Lock()
        self._available = threading.Semaphore(max_connections)

    def acquire(self):
        """获取连接"""
        self._available.acquire()
        with self._lock:
            self.current += 1

    def release(self):
        """释放连接"""
        with self._lock:
            self.current -= 1
        self._available.release()

    def stats(self) -> Dict:
        """获取统计"""
        with self._lock:
            return {
                "current": self.current,
                "max": self.max_connections,
                "available": self.max_connections - self.current,
            }


class AdaptiveRateLimiter:
    """自适应速率限制器（类似 Scrapy AutoThrottle）"""

    def __init__(
        self,
        initial_delay: float = 1.0,
        min_delay: float = 0.1,
        max_delay: float = 60.0,
        target_response_time: float = 2.0,
    ):
        self.initial_delay = initial_delay
        self.min_delay = min_delay
        self.max_delay = max_delay
        self.target_response_time = target_response_time

        self.domain_delays: Dict[str, float] = {}
        self.domain_last_request: Dict[str, float] = {}
        self._lock = threading.Lock()

    def wait(self, url: str):
        """等待"""
        domain = self._extract_domain(url)

        with self._lock:
            delay = self.domain_delays.get(domain, self.initial_delay)
            last_request = self.domain_last_request.get(domain, 0)

            # 等待域延迟
            elapsed = time.time() - last_request
            if elapsed < delay:
                time.sleep(delay - elapsed)

            self.domain_last_request[domain] = time.time()

    def adjust(self, url: str, response_time: float, status_code: int):
        """调整延迟"""
        domain = self._extract_domain(url)

        with self._lock:
            current_delay = self.domain_delays.get(domain, self.initial_delay)
            new_delay = current_delay

            # 根据响应时间调整
            if response_time < self.target_response_time:
                new_delay = current_delay * 0.9
            elif response_time > self.target_response_time * 2:
                new_delay = current_delay * 1.5

            # 根据状态码调整
            if status_code in (429, 503):
                new_delay = current_delay * 2.0
            elif status_code >= 500:
                new_delay = current_delay * 1.2

            # 应用限制
            new_delay = max(self.min_delay, min(self.max_delay, new_delay))

            if abs(new_delay - current_delay) > 0.1:
                self.domain_delays[domain] = new_delay

    def _extract_domain(self, url: str) -> str:
        """提取域名"""
        from urllib.parse import urlparse

        parsed = urlparse(url)
        return parsed.netloc


class ContentFingerprinter:
    """内容指纹（去重）"""

    def __init__(self):
        self.seen_hashes = set()
        self._lock = threading.Lock()

    def generate(self, content: str) -> str:
        import hashlib

        return hashlib.md5(content.encode()).hexdigest()

    def is_duplicate(self, content: str) -> bool:
        """检查是否重复"""
        fingerprint = self.generate(content)

        with self._lock:
            if fingerprint in self.seen_hashes:
                return True
            self.seen_hashes.add(fingerprint)
            return False

    def clear(self):
        """清空"""
        with self._lock:
            self.seen_hashes.clear()
